Begin Profiling Rework to move towards Modularity

maiadegraaf opened the pull request:
_Had to open a new PR but it is the same as #10935._

This PR introduces a new pragma `custom_profiling_settings` which accepts a path to a JSON file.  The file can be used to turn on or off profiling settings when the profiling mode is set to JSON (`PRAGMA enable_profiling = 'json';`).  If the pragma is not set, all profiling information will be included.

Currently, the settings are as follows:

- **CPU Timing**: The total CPU time spent by that operator and its children.
- **Operator Cardinality**: The number of rows returned by that operator
- **Operator Timing**: The time spent by that operator
- **Extra Info**: Any other information collected during the operation


An example of a custom profiling settings JSON file:
```sql
PRAGMA custom_profiling_settings='/path/to/settings.json'
```
```json
{
  "cpu_time":"true",
  "operator_cardinality":"true",
}
```

----
The plan is to expand this feature to be able to include more settings in the future, as well as a new function in the C API that returns an object with the profiling information.

maiadegraaf committed:
be able to set CPU_time, operation_timing, and operation_cardinality on/off

maiadegraaf committed:
be able to set CPU_time, operation_timing, and operation_cardinality on/off

maiadegraaf committed:
parse file in settings instead of query profiler

maiadegraaf committed:
parse file in settings instead of query profiler

maiadegraaf committed:
add TreeNodeSettings

maiadegraaf committed:
add TreeNodeSettings

maiadegraaf committed:
Add extra info setting

maiadegraaf committed:
Add extra info setting

maiadegraaf committed:
Add extra info setting

maiadegraaf committed:
Add extra info setting

maiadegraaf committed:
Add '__TEST_DIR__'

maiadegraaf committed:
Add '__TEST_DIR__'

maiadegraaf committed:
ff

maiadegraaf committed:
ff

maiadegraaf committed:
nits, move tree_node_settings to own file, and move setting to OperatorProfiler.

maiadegraaf committed:
nits, move tree_node_settings to own file, and move setting to OperatorProfiler.

maiadegraaf committed:
Rework TreeNodeSettings

maiadegraaf committed:
Rework TreeNodeSettings

maiadegraaf committed:
Fix extra info

maiadegraaf committed:
Fix extra info

maiadegraaf committed:
ff

maiadegraaf committed:
ff

Mause committed:
build: restore tarball build support

maiadegraaf committed:
generate enum

maiadegraaf committed:
generate enum

maiadegraaf committed:
rm value getters and setters

maiadegraaf committed:
rm value getters and setters

maiadegraaf committed:
ff

maiadegraaf committed:
ff

maiadegraaf committed:
Merge branch 'main' into metrics

maiadegraaf committed:
fix for failing test

maiadegraaf committed:
fix for failing test

maiadegraaf committed:
add custom_profiling_settings to test

maiadegraaf committed:
add custom_profiling_settings to test

maiadegraaf committed:
add build directory

maiadegraaf committed:
add build directory

maiadegraaf committed:
put back test and print non string items as not strings

maiadegraaf committed:
put back test and print non string items as not strings

maiadegraaf committed:
ff

maiadegraaf committed:
ff

maiadegraaf committed:
add __WORKING_DIRECTORY__

maiadegraaf committed:
add __WORKING_DIRECTORY__

maiadegraaf committed:
add mo includes

maiadegraaf committed:
add mo includes

maiadegraaf committed:
add cpu to each child

maiadegraaf committed:
add cpu to each child

maiadegraaf committed:
add include

maiadegraaf committed:
add include

maiadegraaf committed:
add hash>?

maiadegraaf committed:
add hash>?

maiadegraaf committed:
idkkkk

maiadegraaf committed:
idkkkk

maiadegraaf committed:
add hash?

maiadegraaf committed:
add hash?

maiadegraaf committed:
ff

maiadegraaf committed:
ff

maiadegraaf committed:
add rowsort

maiadegraaf committed:
add rowsort

maiadegraaf committed:
put back some stuff

maiadegraaf committed:
Merge remote-tracking branch 'origin/metrics' into metrics

# Conflicts:
#	src/include/duckdb/main/query_profiler.hpp

maiadegraaf committed:
add &

maiadegraaf committed:
make const string

maiadegraaf committed:
ff

maiadegraaf committed:
nits and renaming

maiadegraaf committed:
int -> idx_t

maiadegraaf committed:
add more tests

maiadegraaf committed:
Merge remote-tracking branch 'origin/feature' into metrics

# Conflicts:
#	src/main/query_profiler.cpp

maiadegraaf committed:
add back default settings

maiadegraaf committed:
int -> idx_t and add DefaultSettings()

maiadegraaf committed:
generate enums

maiadegraaf committed:
add more tests

taniabogatsch commented on src/include/duckdb/main/profiling_info.hpp:
```diff
@@ -0,0 +1,83 @@
+	}
+};
+
+typedef unordered_set<MetricsType, MetricsTypeHashFunction> SettingsSet;
```
Nit: can we rename this to `ProfilerSettings`?

taniabogatsch commented on src/include/duckdb/main/profiling_info.hpp:
```diff
@@ -0,0 +1,83 @@
+
+class ProfilingInfo {
+public:
+	// map of metrics with their values; only enabled metrics are present in the map
```
Nit: no longer a `map`.

taniabogatsch commented on src/include/duckdb/main/profiling_info.hpp:
```diff
@@ -0,0 +1,83 @@
+	ProfilingInfo &operator=(ProfilingInfo const &) = default;
+
+public:
+	// set the metrics map
```
Nit: no longer a `map`. Parameter can be `const`?

taniabogatsch commented on src/include/duckdb/main/profiling_info.hpp:
```diff
@@ -0,0 +1,83 @@
+public:
+	// set the metrics map
+	void SetSettings(SettingsSet &n_settings);
+	// get the metrics map
```
Nit: no longer a `map`. Should this return a `const` reference?

taniabogatsch commented on src/include/duckdb/main/settings.hpp:
```diff
@@ -267,6 +267,16 @@ struct EnableProfilingSetting {
+struct CustomProfilingSettings {
+	static constexpr const char *Name = "custom_profiling_settings";
+	static constexpr const char *Description =
+	    "Accepts a path to a JSON file which is used to customize the profiling statistics";
```
`"Accepts a path to a JSON file enabling custom metrics."

taniabogatsch commented on src/include/duckdb/main/query_profiler.hpp:
```diff
@@ -80,46 +82,61 @@ struct ExpressionExecutorInfo {
 	//! A vector of Expression Executor Info
 	vector<unique_ptr<ExpressionExecutorInfo>> executors_info;
+
+	void addTime(double n_time) {
```
Uppercase.

taniabogatsch commented on src/include/duckdb/main/query_profiler.hpp:
```diff
@@ -80,46 +82,61 @@ struct ExpressionExecutorInfo {
+		this->time += n_time;
+	}
+
+	void addElements(idx_t n_elements) {
```
Uppercase.

taniabogatsch commented on src/include/duckdb/main/query_profiler.hpp:
```diff
@@ -128,22 +145,29 @@ class QueryProfiler {
 		string name;
-		string extra_info;
-		OperatorInformation info;
+		ProfilingInfo settings;
```
Nit: this includes both the settings and the metrics now.

taniabogatsch commented on src/include/duckdb/main/query_profiler.hpp:
```diff
@@ -128,22 +145,29 @@ class QueryProfiler {
 
 private:
-	unique_ptr<TreeNode> CreateTree(const PhysicalOperator &root, idx_t depth = 0);
+	unique_ptr<QueryProfiler::TreeNode> CreateTree(const PhysicalOperator &root, SettingsSet settings, idx_t depth = 0);
```
This cannot be `<TreeNode>` any longer?

taniabogatsch commented on src/main/query_profiler.cpp:
```diff
@@ -126,33 +126,59 @@ void QueryProfiler::StartExplainAnalyze() {
 	// EXPLAIN ANALYSE should not be outputted by the profiler
 	if (IsEnabled() && !is_explain_analyze) {
-		string query_info = ToString();
+		// add an outer node for the query
```
Nit: no longer a node.

taniabogatsch reviewed (approved):
Looks neat now! I only had a few nits. :)

maiadegraaf committed:
nits

maiadegraaf committed:
Merge remote-tracking branch 'origin/feature' into metrics_rework

maiadegraaf committed:
add check for empty file

maiadegraaf committed:
rename file

maiadegraaf committed:
print line

maiadegraaf committed:
alter file so extra wrong

maiadegraaf committed:
add extra exception

maiadegraaf committed:
revert LinuxRelease.yml

maiadegraaf committed:
ff

maiadegraaf committed:
ff

maiadegraaf committed:
Merge remote-tracking branch 'origin/feature' into metrics_rework

maiadegraaf committed:
fix force reload

maiadegraaf committed:
Merge remote-tracking branch 'origin/feature' into metrics_rework

# Conflicts:
#	src/common/enum_util.cpp
#	src/common/tree_renderer.cpp
#	src/include/duckdb/main/query_profiler.hpp
#	src/main/config.cpp
#	src/main/query_profiler.cpp
#	test/api/test_reset.cpp

maiadegraaf committed:
fix merge conflicts

maiadegraaf committed:
ff

maiadegraaf committed:
run generate files

maiadegraaf committed:
fix explain analyze test

maiadegraaf committed:
clang tidy

maiadegraaf committed:
ff

maiadegraaf committed:
Change custom profiler settings to accepts a JSON instead of a json file.

maiadegraaf committed:
adjust test

taniabogatsch commented:
Waiting on https://github.com/duckdb/duckdb/pull/11868.
@maiadegraaf, please ping me once this is ready for review!

maiadegraaf committed:
rm JSON extension from test

Maxxen commented:
https://github.com/duckdb/duckdb/pull/11868 has been superseded by https://github.com/duckdb/duckdb/pull/11998 so this should be unblocked now!

maiadegraaf committed:
Merge remote-tracking branch 'refs/remotes/origin/feature' into metrics_rework

maiadegraaf committed:
add json

taniabogatsch commented:
Still waiting on merging `main` into `feature` to unlock the new JSON features.

maiadegraaf committed:
Merge remote-tracking branch 'refs/remotes/origin/feature' into metrics_rework

# Conflicts:
#	src/common/enum_util.cpp
#	src/common/string_util.cpp

maiadegraaf committed:
add back metrics enum

maiadegraaf committed:
Merge remote-tracking branch 'refs/remotes/origin/feature' into metrics_rework

maiadegraaf committed:
Merge remote-tracking branch 'refs/remotes/origin/feature' into metrics_rework

taniabogatsch reviewed (approved):
The CI failure seems unrelated.

maiadegraaf commented:
> The CI failure seems unrelated.

Its because I changed the name of certain proponents in the profiling, I updated the script so should work from now on. Just fails in the CI on this PR.
